Add OpenCL GPU implementation

sssarana opened the pull request:
- add dependency on `grid_map_core`;
- fix build with NONFREE algorithms for `humble-devel`;
- Replace Mat objects with UMat to use with OpenCL;

This will automatically use OpenCL GPU if it is available.

matlabbe commented on corelib/src/Features2d.cpp:
```diff
@@ -1140,6 +1145,7 @@ void SURF::parseParameters(const ParametersMap & parameters)
 	{
 #if CV_MAJOR_VERSION < 3
 		_surf = cv::Ptr<CV_SURF>(new CV_SURF(hessianThreshold_, nOctaves_, nOctaveLayers_, extended_, upright_));
+		std::cout << "using cpu" << std::endl;
```
That log could be removed.

matlabbe commented on corelib/src/Features2d.cpp:
```diff
@@ -1172,11 +1178,20 @@ std::vector<cv::KeyPoint> SURF::generateKeypointsImpl(const cv::Mat & image, con
-	{
-		_surf->detect(imgRoi, keypoints, maskRoi);
-	}
+    }
+    else
+    {
+        cv::UMat imgRoiU;
+        imgRoi.copyTo(imgRoiU);
+
+        cv::UMat maskRoiU;
+        if (!maskRoi.empty())
+        {
+            maskRoi.copyTo(maskRoiU);
+        }
+
+        _surf->detect(imgRoiU, keypoints, maskRoiU);
+    }
```
Would we need to do something similar for all detectors? I can help if so.

sssarana replied:
Yeah, all detectors should have similar implementation if we want to use them on GPU. It would be great to have it.

matlabbe commented on corelib/src/MarkerDetector.cpp:
```diff
@@ -246,11 +247,13 @@ std::map<int, MarkerInfo> MarkerDetector::detect(const cv::Mat & image,
-				float d3 = util2d::getDepth(depth, corners[i][2].x*rgbToDepthFactorX, corners[i][2].y*rgbToDepthFactorY, true, 0.02f, true);
-				float d4 = util2d::getDepth(depth, corners[i][3].x*rgbToDepthFactorX, corners[i][3].y*rgbToDepthFactorY, true, 0.02f, true);
+				cv::Mat depthMat;
+				depth.copyTo(depthMat);
```
```suggestion
				depthMat = depth.getMat(cv::ACCESS_READ);

```
copy only if needed

sssarana replied:
The `depthMat = depth.getMat(cv::ACCESS_READ);` caused some lifecycle issues for me earlier, so I wouldn't trust it. The reason for that is the original object that we create a shallow copy from gets deleted, so the `depthMat` gets deleted also.

matlabbe commented on corelib/src/MarkerDetector.cpp:
```diff
@@ -363,9 +366,12 @@ std::map<int, MarkerInfo> MarkerDetector::detect(const cv::Mat & image,
 	{
+		cv::Mat detections_img;
+		imageWithDetections->copyTo(detections_img);
+		UWARN("entering this part");
```
Log to remove.

matlabbe commented on corelib/src/odometry/OdometryF2M.cpp:
```diff
@@ -1263,6 +1266,7 @@ Transform OdometryF2M::computeTransform(
 
 					map_->setWords(words, wordsKpts, transformedPoints, descriptors);
 					addKeyFrame = true;
+					descriptors.release();
```
Is it necessary?

matlabbe commented on package.xml:
```diff
@@ -22,7 +21,7 @@
   <depend>libsqlite3-dev</depend>
   <depend>liboctomap-dev</depend>
-  <!-- <depend>grid_map_core</depend> --> <!-- till this PR is released https://github.com/ANYbotics/grid_map/pull/499 -->
+  <depend>grid_map_core</depend> <!-- till this PR is released https://github.com/ANYbotics/grid_map/pull/499 -->
```
I think we still need to wait for https://github.com/ANYbotics/grid_map/pull/519 to be released, we may change the comment to 
```suggestion
  <!-- <depend>grid_map_core</depend> --> <!-- till this PR is released https://github.com/ANYbotics/grid_map/pull/519 -->
```

sssarana replied:
I created a PR to fix `grid_map_core` here: ANYbotics/grid_map#521, but it didn't get reviewed yet. Though I think you can apply these fixes yourself if you want to test everything properly.

matlabbe commented:
I kinda forgot about `UMat` when OpenCV3 came out, thinking it was already integrated inside `cv::Mat`, though it should be explicitly used. I put some comments above. It looks like for most functions, we would need to convert to `cv::InputArray` / `cv::OutputArray` to avoid unnecessary copies between GPU and CPU. That would be a bigger PR though. I'll try to take time to test at least your changes to see what is the impact on those functions (using OpenCL or not) compared to old code, then merge it for now (and do UMat support for all other detectors later).

sssarana commented:
Thanks for your feedback. I fixed the errors and removed unnecessary logs now, let me know if anything else is needed.
Using the `cv::InputArray`/`cv::OutputArray` is a good idea, it will also make implementation for `UMat` support on other detectors a bit easier. 
Thanks.
